tests/integration/user/user_test.go

```diff
@@ -0,0 +1,243 @@
+// Copyright 2021 PingCAP, Inc. Licensed under Apache-2.0.
+
+package user
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"github.com/gin-gonic/gin"
+	"github.com/joomcode/errorx"
+	"github.com/stretchr/testify/suite"
+	"go.uber.org/fx"
+
+	"github.com/pingcap/tidb-dashboard/pkg/apiserver/info"
+	"github.com/pingcap/tidb-dashboard/pkg/apiserver/user"
+	"github.com/pingcap/tidb-dashboard/pkg/apiserver/user/sqlauth"
+	"github.com/pingcap/tidb-dashboard/pkg/config"
+	"github.com/pingcap/tidb-dashboard/pkg/dbstore"
+	"github.com/pingcap/tidb-dashboard/pkg/httpc"
+	"github.com/pingcap/tidb-dashboard/pkg/pd"
+	"github.com/pingcap/tidb-dashboard/pkg/tidb"
+	"github.com/pingcap/tidb-dashboard/tests/util"
+	"github.com/pingcap/tidb-dashboard/util/featureflag"
+	"github.com/pingcap/tidb-dashboard/util/rest"
+	"github.com/pingcap/tidb-dashboard/util/testutil"
+)
+
+type testUserSuite struct {
+	suite.Suite
+	db          *testutil.TestDB
+	authService *user.AuthService
+	infoService *info.Service
+}
+
+func TestUserSuite(t *testing.T) {
+	db := testutil.OpenTestDB(t)
+	tidbVersion := util.GetTiDBVersion(t, db)
+
+	authService := &user.AuthService{}
+	infoService := &info.Service{}
+
+	app := fx.New(
+		fx.Supply(featureflag.NewRegistry(tidbVersion)),
+		fx.Supply(config.Default()),
+		fx.Provide(
+			httpc.NewHTTPClient,
+			pd.NewEtcdClient,
+			tidb.NewTiDBClient,
+			user.NewAuthService,
+			dbstore.NewDBStore,
+			info.NewService,
+		),
+		sqlauth.Module,
+		fx.Populate(&authService),
+		fx.Populate(&infoService),
+	)
+	ctx := context.Background()
+	_ = app.Start(ctx)
+
+	suite.Run(t, &testUserSuite{
+		db:          db,
+		authService: authService,
+		infoService: infoService,
+	})
+
+	_ = app.Stop(ctx)
+}
+
+func (s *testUserSuite) supportNonRootLogin() bool {
+	return s.authService.FeatureFlagNonRootLogin.IsSupported()
+}
+
+func genReq(method, uri string, param map[string]interface{}) (*gin.Context, *httptest.ResponseRecorder) {
+	var jsonByte []byte
+	if param != nil {
+		jsonByte, _ = json.Marshal(param)
+	}
+
+	w := httptest.NewRecorder()
+	c, _ := gin.CreateTestContext(w)
+	c.Request, _ = http.NewRequest(method, uri, bytes.NewReader(jsonByte))
+	return c, w
+}
```

Comment:
Maybe we can just use `CreateTestContext` to test. Then most cases we don't need to validate *httptest.ResponseRecorder.

gin_test_helper.go
```
func TestReqWithHandlers(req *http.Request, handlers ...gin.HandlerFunc) (*gin.Context, *httptest.ResponseRecorder) {
	w := httptest.NewRecorder()
	c, e := gin.CreateTestContext(w)
	c.Request = req
	e.Any(req.RequestURI, handlers...)
	e.HandleContext(c)
	return c, w
}
```
xxx_test.go
```
req, _ := http.NewRequest("POST", "/user/login", bytes.NewReader(jsonByte))
ctx, _ := TestReqWithHandlers(req, s.FeatureTopSQL.VersionGuard(), provideTestDB, s.authService.LoginHandler)

...
// s.Require().True(errorx.IsOfType(ctx.Errors.Last().Err, rest.ErrBadRequest))
// s.Require().Equal(401, ctx.Writer.Status())
```

Reply:
Let me have a look

Reply:
it works, and `TestReqWithHandlers` can be refined to:

```go
	w := httptest.NewRecorder()
	c, e := gin.CreateTestContext(w)
	c.Request = req
	e.Handle(req.Method, req.URL.Path, handlers...)
	e.HandleContext(c)
	return c, w
```

`req.RequestURI` is empty.